### PR TITLE
6051 bulk export security errors

### PR DESCRIPTION
**Pull request by tadgh**
Refactor rules for bulk export to be safer.

### CODE REVIEW

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImplTest.java [hunk 1]
```diff
@@ -42,33 +45,210 @@ public void testDenyBulkRequestWithInvalidResourcesTypes() {
...
+				AuthorizationInterceptor.Verdict verdict = myRule.applyRule(myOperation, myRequestDetails, null, null, null, myRuleApplier, myFlags, myPointcut);
+
+				assertAllow(verdict);
+
+			}
+
+		}
+
+		@Nested
+		class RuleSystemStyle {
```
Comment (fil512) [inline hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImplTest.java:153]: this is a nice way to organize the tests!

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java [hunk 2]
```diff
@@ -70,115 +70,85 @@ public AuthorizationInterceptor.Verdict applyRule(
 			return null;
 		}
 
-		BulkExportJobParameters options = (BulkExportJobParameters)
+		BulkExportJobParameters inboundBulkExportRequestOptions = (BulkExportJobParameters)
 				theRequestDetails.getAttribute(AuthorizationInterceptor.REQUEST_ATTRIBUTE_BULK_DATA_EXPORT_OPTIONS);
-
-		if (!myWantAnyStyle && options.getExportStyle() != myWantExportStyle) {
+		// if style doesn't match - abstain
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java:75]: question: is abstain different from deny?  (i.e. no decision leave it up to caller / subsequent rules how to interpret that?)

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java [hunk 3]
```diff
@@ -70,115 +70,85 @@ public AuthorizationInterceptor.Verdict applyRule(
...
-							theRequestDetails,
-							theInputResource,
-							theInputResourceId,
-							theOutputResource,
-							theRuleApplier);
+		// Instance level, or filtered type level
+		if (isNotEmpty(myPatientIds)) {
+			// If bulk export options defines no patient IDs, return null.
+			if (inboundBulkExportRequestOptions.getPatientIds().isEmpty()) {
+				return null;
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java:132]: nope this is intentional. If they provided permissioned rules, and dont provide a filtered request, it gets rejected (e.g. null). The magic is on the CDR side where we auto-narrow if the caller omits the patient IDs.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java (hunk 2 above)
Comment (tadgh) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java:75]: You got it. Abstaining is saying "I dont approve, maybe somebody later will". IF it hits the end of the chain with no explicit approval, we deny.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBuilder.java [hunk 4]
```diff
@@ -895,6 +895,23 @@ public IAuthRuleBuilderRuleBulkExportWithTarget groupExportOnGroup(@Nonnull Stri
...
+				if (ruleBulkExport == null) {
+					RuleBulkExportImpl rule = new RuleBulkExportImpl(myRuleName);
+					rule.setMode(myRuleMode);
+					ruleBulkExport = rule;
+				}
+				ruleBulkExport.setAppliesToPatientExportAllPatients();
+
+				// prevent duplicate rules being added
+				if (!myRules.contains(ruleBulkExport)) {
+					myRules.add(ruleBulkExport);
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBuilder.java:909]: I don't want to touch a structural change like that this late in the game, will open a new ticket.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java [hunk 5]
```diff
@@ -70,115 +70,85 @@ public AuthorizationInterceptor.Verdict applyRule(
...
-		if (!myWantAnyStyle && options.getExportStyle() != myWantExportStyle) {
+		// if style doesn't match - abstain
+		if (!myWantAnyStyle && inboundBulkExportRequestOptions.getExportStyle() != myWantExportStyle) {
 			return null;
 		}
 
+		// Do we only authorize some types?  If so, make sure requested types are a subset
 		if (isNotEmpty(myResourceTypes)) {
-			if (isEmpty(options.getResourceTypes())) {
+			if (isEmpty(inboundBulkExportRequestOptions.getResourceTypes())) {
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java:82]: I don't know what that means. Like, should we warn? I would say no. If a rule has explicit resource types, and the incoming request omits them, that will imply to bulk export to auto-inflate to all possible types, which it isnt allowed to do, so we reject.

**Pull request merged**